src/dbami/helpers/apply_migrations.py

```diff
@@ -0,0 +1,117 @@
+import logging
+import time
+from typing import Literal, Optional, Sequence
+
+import asyncpg
+from buildpg import V, render
+
+from dbami.db import DB
+
+logger = logging.getLogger(__name__)
+
+
+async def grant_connect_privileges(conn: asyncpg.Connection, *role_names: str) -> None:
+    for role_name in role_names:
+        logger.warning("Granting connect privileges for role '%s'", role_name)
+        q, p = render(
+            """
+                DO $_$
+                    BEGIN
+                        EXECUTE FORMAT('GRANT CONNECT on database %s TO :role',
+                        CURRENT_DATABASE());
+                    END
+                $_$;
+            """,
+            role=V(role_name),
+        )
+        await conn.execute(q, *p)
+
+
+async def revoke_connect_privileges(conn: asyncpg.Connection, *role_names: str) -> None:
+    for role_name in role_names:
+        logger.warning("Revoking connect privileges for role '%s'", role_name)
+        q, p = render(
+            """
+                DO $_$
+                    BEGIN
+                        EXECUTE FORMAT('REVOKE CONNECT on database %s FROM :role',
+                        CURRENT_DATABASE());
+                    END
+                $_$;
+            """,
+            role=V(role_name),
+        )
+        await conn.execute(q, *p)
+
+
+async def active_connections_exist(conn: asyncpg.Connection) -> bool:
+    return bool(
+        await conn.fetchval(
+            """
+        SELECT EXISTS(
+            SELECT * FROM pg_stat_activity
+            WHERE
+                datname = current_database()
+                AND pid != pg_backend_pid()
```

Comment:
Need to change this to check for connections not of this user, nit just not this PID.

Reply:
I went with checking that the users/roles that the app uses are disconnected. I didn't want this to hang because I had a idle-ish pgAdmin4 connection open as postgres role to hang migration. 

```sql
SELECT * FROM pg_stat_activity
WHERE datname = current_database()
AND usename = any(ARRAY['read_app_user', 'write_app_user'])
```